File: examples/agents/python/agent.py
#!/usr/bin/env python3
"""Minimal vAuth presence agent using dbus-next.

This console example cancels secret prompts. A graphical agent may call
submit_secret() with a mutable buffer populated by a protected input widget,
then clear that widget and buffer immediately.
"""

import asyncio
import os
from typing import Optional

from dbus_next import BusType, Message, MessageType
from dbus_next.aio import MessageBus


SERVICE = "org.lamellix.vAuth"
PATH = "/org/lamellix/vAuth"
INTERFACE = "org.lamellix.vAuth.UserInteraction1"
START_STATES = {"presence_required", "verification_started"}
TERMINAL_STATES = {
    "presence_approved",
    "presence_denied",
    "verification_succeeded",
    "verification_failed",
    "cancelled",
    "timed_out",
}
KNOWN_STATES = START_STATES | TERMINAL_STATES | {
    "verification_information",
    "verification_error",
    "secret_required",
}


class Agent:
    def __init__(self) -> None:
        self.bus: Optional[MessageBus] = None
        self.interface = None
        self.generation = 0
        self.active_request: Optional[int] = None

    async def connect(self) -> None:
        self.bus = await MessageBus(bus_type=BusType.SYSTEM).connect()
        introspection = await self.bus.introspect(SERVICE, PATH)
        proxy = self.bus.get_proxy_object(SERVICE, PATH, introspection)
        self.interface = proxy.get_interface(INTERFACE)

        # Subscribe before registration so an early directed event is queued.
        self.interface.on_state_changed(self.on_state_changed)
        self.generation = await self.interface.call_register_agent()
        if self.generation == 0:
            raise RuntimeError("daemon returned generation zero")
        print(f"registered generation {self.generation}")

    def on_state_changed(
        self,
        generation: int,
        request_id: int,
        prompt_id: int,
        state: str,
        operation: str,
        relying_party_id: str,
        message: str,
    ) -> None:
        asyncio.create_task(
            self.handle_state(
                generation,
                request_id,
                prompt_id,
                state,
                operation,
                relying_party_id,
                message,
            )
        )

    async def handle_state(
        self,
        generation: int,
        request_id: int,
        prompt_id: int,
        state: str,
        operation: str,
        relying_party_id: str,
        message: str,
    ) -> None:
        if (
            generation != self.generation
            or request_id == 0
            or state not in KNOWN_STATES
        ):
            return
        if (state == "secret_required") != (prompt_id != 0):
            return
        if state in START_STATES:
            if self.active_request not in (None, request_id):
                return
            self.active_request = request_id
        elif self.active_request != request_id:
            return

        print(f"{state}: {operation} for RP {relying_party_id}")
        if message:
            print(message)

        if state == "presence_required":
            answer = await asyncio.to_thread(
                input, "Approve? [y]es/[n]o/[c]ancel: "
            )
            try:
                if answer.strip().lower().startswith("c"):
                    await self.interface.call_cancel_interaction(
                        self.generation, request_id
                    )
                else:
                    await self.interface.call_respond_to_presence(
                        self.generation,
                        request_id,
                        answer.strip().lower().startswith("y"),
                    )
            except Exception as error:
                print(f"presence reply failed: {error}")
        elif state == "secret_required":
            # Console strings cannot be reliably erased. A real UI should call
            # submit_secret() with bytes from a protected input widget.
            try:
                await self.interface.call_cancel_interaction(
                    self.generation, request_id
                )
            except Exception as error:
                print(f"secret cancellation failed: {error}")

        if state in TERMINAL_STATES:
            self.active_request = None

    async def submit_secret(
        self, request_id: int, prompt_id: int, secret: bytearray
    ) -> None:
        """Submit and erase a mutable verification-secret buffer."""
        read_fd = -1
        write_fd = -1
        try:
            if (
                self.bus is None
                or prompt_id == 0
                or self.active_request != request_id
            ):
                raise RuntimeError("interaction is not active")
            if len(secret) > 1024 or 0 in secret:
                raise ValueError(
                    "secret must be at most 1024 bytes without NUL"
                )

            read_fd, write_fd = os.pipe2(os.O_CLOEXEC)
            written = 0
            while written < len(secret):
                written += os.write(write_fd, memoryview(secret)[written:])
            os.close(write_fd)
            write_fd = -1

            # A D-Bus 'h' value indexes the ancillary unix_fds array.
            reply = await self.bus.call(
                Message(
                    destination=SERVICE,
                    path=PATH,
                    interface=INTERFACE,
                    member="SubmitSecret",
                    signature="ttth",
                    body=[self.generation, request_id, prompt_id, 0],
                    unix_fds=[read_fd],
                )
            )
            if reply.message_type == MessageType.ERROR:
                detail = str(reply.body[0]) if reply.body else "D-Bus error"
                raise RuntimeError(f"{reply.error_name}: {detail}")
        finally:
            if write_fd >= 0:
                os.close(write_fd)
            if read_fd >= 0:
                os.close(read_fd)
            secret[:] = b"\0" * len(secret)
            secret.clear()

    async def close(self) -> None:
        if self.interface is not None and self.generation != 0:
            try:
                await self.interface.call_unregister_agent()
            except Exception:
                pass
        if self.bus is not None:
            self.bus.disconnect()


async def main() -> None:
    agent = Agent()
    await agent.connect()
    try:
        await asyncio.Future()
    finally:
        await agent.close()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        pass

File: examples/agents/cpp/agent.cpp
#include <sdbus-c++/sdbus-c++.h>

#include <algorithm>
#include <array>
#include <cerrno>
#include <condition_variable>
#include <cstdint>
#include <deque>
#include <fcntl.h>
#include <iostream>
#include <memory>
#include <mutex>
#include <span>
#include <stdexcept>
#include <string>
#include <string.h>
#include <string_view>
#include <system_error>
#include <unistd.h>
#include <utility>
#include <vector>

namespace {

constexpr char SERVICE[] = "org.lamellix.vAuth";
constexpr char PATH[] = "/org/lamellix/vAuth";
constexpr char INTERFACE[] = "org.lamellix.vAuth.UserInteraction1";

struct Event {
    uint64_t generation;
    uint64_t requestId;
    uint64_t promptId;
    std::string state;
    std::string operation;
    std::string relyingPartyId;
    std::string message;
};

class UniqueFd {
public:
    explicit UniqueFd(int fd = -1) noexcept : fd_(fd) {}
    ~UniqueFd() {
        if(fd_ >= 0)
            static_cast<void>(close(fd_));
    }
    UniqueFd(const UniqueFd&) = delete;
    UniqueFd& operator=(const UniqueFd&) = delete;
    UniqueFd(UniqueFd&& other) noexcept
        : fd_(std::exchange(other.fd_, -1)) {}
    [[nodiscard]] int get() const noexcept { return fd_; }
    void reset() noexcept {
        if(fd_ >= 0)
            static_cast<void>(close(std::exchange(fd_, -1)));
    }

private:
    int fd_;
};

void write_all(int fd, std::span<const uint8_t> bytes) {
    std::size_t offset = 0;
    while(offset < bytes.size()) {
        const ssize_t count = write(
            fd, bytes.data() + offset, bytes.size() - offset
        );
        if(count > 0) {
            offset += static_cast<std::size_t>(count);
        } else if(count < 0 && errno == EINTR) {
            continue;
        } else {
            throw std::system_error(
                count < 0 ? errno : EIO,
                std::generic_category(),
                "write secret pipe"
            );
        }
    }
}

bool is_start(std::string_view state) {
    return state == "presence_required" || state == "verification_started";
}

bool is_terminal(std::string_view state) {
    return
        state == "presence_approved" ||
        state == "presence_denied" ||
        state == "verification_succeeded" ||
        state == "verification_failed" ||
        state == "cancelled" ||
        state == "timed_out";
}

bool is_known(std::string_view state) {
    return
        is_start(state) ||
        is_terminal(state) ||
        state == "verification_information" ||
        state == "verification_error" ||
        state == "secret_required";
}

void clear_secret(std::vector<uint8_t>& secret) noexcept {
    if(!secret.empty())
        explicit_bzero(secret.data(), secret.size());
    secret.clear();
}

class SecretClearGuard {
public:
    explicit SecretClearGuard(std::vector<uint8_t>& secret) noexcept
        : secret_(secret) {}
    ~SecretClearGuard() { clear_secret(secret_); }
    SecretClearGuard(const SecretClearGuard&) = delete;
    SecretClearGuard& operator=(const SecretClearGuard&) = delete;

private:
    std::vector<uint8_t>& secret_;
};

class Agent {
public:
    Agent()
        : connection_(sdbus::createSystemBusConnection()),
          proxy_(sdbus::createProxy(
              *connection_,
              sdbus::ServiceName{SERVICE},
              sdbus::ObjectPath{PATH}
          )) {
        stateSlot_ = proxy_->uponSignal("StateChanged")
            .onInterface(INTERFACE)
            .call(
                [this](
                    uint64_t generation,
                    uint64_t request_id,
                    uint64_t prompt_id,
                    const std::string& state,
                    const std::string& operation,
                    const std::string& relying_party_id,
                    const std::string& message
                ) {
                    std::lock_guard lock(mutex_);
                    events_.push_back({
                        generation,
                        request_id,
                        prompt_id,
                        state,
                        operation,
                        relying_party_id,
                        message
                    });
                    changed_.notify_one();
                },
                sdbus::return_slot
            );

        proxy_->callMethod("RegisterAgent")
            .onInterface(INTERFACE)
            .storeResultsTo(generation_);
        if(generation_ == 0)
            throw std::runtime_error("daemon returned generation zero");
        connection_->enterEventLoopAsync();
        std::cout << "registered generation " << generation_ << '\n';
    }

    ~Agent() {
        try {
            proxy_->callMethod("UnregisterAgent").onInterface(INTERFACE);
        } catch(...) {
        }
        try {
            connection_->leaveEventLoop();
        } catch(...) {
        }
    }

    Agent(const Agent&) = delete;
    Agent& operator=(const Agent&) = delete;

    void run() {
        while(true) {
            Event event = next_event();
            if(
                event.generation != generation_ ||
                event.requestId == 0 ||
                !is_known(event.state)
            ) {
                continue;
            }
            if(
                (event.state == "secret_required") !=
                (event.promptId != 0)
            ) {
                continue;
            }
            if(is_start(event.state)) {
                if(activeRequest_ != 0 && activeRequest_ != event.requestId)
                    continue;
                activeRequest_ = event.requestId;
            } else if(activeRequest_ != event.requestId) {
                continue;
            }

            std::cout << event.state << ": " << event.operation
                      << " for RP " << event.relyingPartyId << '\n';
            if(!event.message.empty())
                std::cout << event.message << '\n';

            if(event.state == "presence_required") {
                std::cout << "Approve? [y]es/[n]o/[c]ancel: " << std::flush;
                std::string answer;
                if(!std::getline(std::cin, answer) ||
                    (!answer.empty() && answer.front() == 'c')) {
                    cancel(event.requestId);
                } else {
                    const bool approved =
                        !answer.empty() && answer.front() == 'y';
                    respond_to_presence(event.requestId, approved);
                }
            } else if(event.state == "secret_required") {
                // A console string cannot be reliably erased. A real UI should
                // call submit_secret() with a protected mutable buffer.
                cancel(event.requestId);
            }

            if(is_terminal(event.state))
                activeRequest_ = 0;
        }
    }

    void submit_secret(
        uint64_t request_id,
        uint64_t prompt_id,
        std::vector<uint8_t>& secret
    ) {
        SecretClearGuard clearOnExit(secret);
        if(request_id == 0 || prompt_id == 0 || request_id != activeRequest_)
            throw std::runtime_error("interaction is not active");
        if(
            secret.size() > 1024 ||
            std::ranges::find(secret, uint8_t{0}) != secret.end()
        ) {
            throw std::invalid_argument(
                "secret must be at most 1024 bytes without NUL"
            );
        }

        std::array<int, 2> descriptors{};
        if(pipe2(descriptors.data(), O_CLOEXEC) != 0)
            throw std::system_error(errno, std::generic_category(), "pipe2");
        UniqueFd readEnd(descriptors[0]);
        UniqueFd writeEnd(descriptors[1]);

        write_all(writeEnd.get(), secret);
        writeEnd.reset();
        proxy_->callMethod("SubmitSecret")
            .onInterface(INTERFACE)
            .withArguments(
                generation_,
                request_id,
                prompt_id,
                sdbus::UnixFd{readEnd.get()}
            );
    }

private:
    Event next_event() {
        std::unique_lock lock(mutex_);
        changed_.wait(lock, [this] { return !events_.empty(); });
        Event event = std::move(events_.front());
        events_.pop_front();
        return event;
    }

    void respond_to_presence(uint64_t request_id, bool approved) {
        proxy_->callMethod("RespondToPresence")
            .onInterface(INTERFACE)
            .withArguments(generation_, request_id, approved);
    }

    void cancel(uint64_t request_id) {
        proxy_->callMethod("CancelInteraction")
            .onInterface(INTERFACE)
            .withArguments(generation_, request_id);
    }

    std::unique_ptr<sdbus::IConnection> connection_;
    std::unique_ptr<sdbus::IProxy> proxy_;
    sdbus::Slot stateSlot_;
    uint64_t generation_ = 0;
    uint64_t activeRequest_ = 0;
    std::mutex mutex_;
    std::condition_variable changed_;
    std::deque<Event> events_;
};

} // namespace

int main() {
    try {
        Agent agent;
        agent.run();
    } catch(const std::exception& error) {
        std::cerr << "vAuth agent example: " << error.what() << '\n';
        return 1;
    }
}
